### PR TITLE
Merge release into master

### PR DESCRIPTION
**Pull request by florianhumblot**
Currently the release branch consists of a buffered and an unbuffered driver for the SSD1306 display.

### CODE REVIEW

File: code/headers/ssd1306.hpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::display_all_on_resume,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::normal_display,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::display_on};
+
+    class ssd1306_i2c_c {
+    protected:
+        /// current cursor location in the controller
```
Comment (itzandroidtab) [inline code/headers/ssd1306.hpp]: this comment is probably with the wrong variable. (pretty sure a i2c_bus is not a cursor location)

File: code/src/ssd1306_oled_unbuffered.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        const uint8_t d = (background == hwlib::white) ? 0xFF : 0x00;
+        command(ssd1306_commands::column_addr, 0, 127);
+        command(ssd1306_commands::page_addr, 0, 7);
+        uint8_t data[sizeof(buffer) + 1] = {0};
+        data[0] = ssd1306_data_prefix;
+        for (uint_fast16_t x = 1; x < sizeof(data); ++x) {
+            data[x] = d;
+        }
+        bus.write(address, data, sizeof(data));
+        cursor = wsize;
```
Comment (itzandroidtab) [inline code/src/ssd1306_oled_unbuffered.cpp]: Is this correct? After writing all the data the cursor falls back to 0,0 right?

File: code/main.cpp [hunk 3]
```diff
@@ -1,9 +1,31 @@
-#include "hwlib.hpp"
-
+#include <hwlib.hpp>
```
Comment (itzandroidtab) [inline code/main.cpp]: Maybe we should move this file to an examples directory. Otherwise this file will change every time we want to push a different screen

File: code/headers/ssd1306_oled_buffered.hpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#include <i2c_bus.hpp>
+#include <ssd1306.hpp>
+
+namespace r2d2::display {
+    /**
+     * SSD1306 buffered interface for an oled
+     * Implements hwlib::window to easily use text and drawing functions that are already implemented.
+     * Extends from r2d2::display::ssd1306_i2c_c 
+     */
+    class ssd1306_oled_buffered : public r2d2::display::ssd1306_i2c_c,
```
Comment (itzandroidtab) [inline code/headers/ssd1306_oled_buffered.hpp]: The naming scheme of this class is not following the style guide

File: code/headers/ssd1306_oled_unbuffered.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#include <i2c_bus.hpp>
+#include <ssd1306.hpp>
+
+namespace r2d2::display {
+    /**
+     * SSD1306 unbuffered interface for an oled
+     * Implements hwlib::window to easily use text and drawing functions that are already implemented.
+     * Extends from r2d2::display::ssd1306_i2c_c 
+     */
+    class ssd1306_oled_unbuffered : public r2d2::display::ssd1306_i2c_c,
```
Comment (itzandroidtab) [inline code/headers/ssd1306_oled_unbuffered.hpp]: The naming scheme of this class is not following the style guide

File: code/headers/ssd1306_oled_buffered.hpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
+#include <hwlib.hpp>
```
Comment (itzandroidtab) [inline code/headers/ssd1306_oled_buffered.hpp]: You don't have include guards in your header

File: code/headers/ssd1306_oled_unbuffered.hpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
+#include <hwlib.hpp>
```
Comment (itzandroidtab) [inline code/headers/ssd1306_oled_unbuffered.hpp]: You don't have include guards in your header

File: code/headers/ssd1306.hpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    /// value to send over i2c before data
+    constexpr uint8_t ssd1306_data_prefix = 0x40;
+
+    /// SSD1306 chip initialization
+    constexpr const uint8_t ssd1306_initialization[] = {
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::display_off,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_display_clock_div,
+        0x80,
```
Comment (LRstudentHU) [inline code/headers/ssd1306.hpp]: What are these magic numbers between the different commands?

File: code/headers/ssd1306.hpp [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        uint8_t address;
+        /// The current cursor location in the controller
+        hwlib::xy cursor;
+
+    public:
+        // construct by providing the i2c channel
+        ssd1306_i2c_c(r2d2::i2c::i2c_bus_c &bus, const uint8_t &address);
+
+        /// send a command without data
+        void command(ssd1306_commands c);
```
Comment (LRstudentHU) [inline code/headers/ssd1306.hpp]: These function names are a bit inconsistent. Maybe the following is more logical?
 - `write_command`
- `write_data`
- `write_pixel`

File: code/headers/ssd1306_oled_buffered.hpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+         * The size of the display
+         * This is needed even though hwlib::window keeps it's own size to create the buffer at compile time.
+         */ 
+        static auto constexpr wsize = hwlib::xy(128, 64);
+
+        /**
+         * The buffer with the pixel data
+         * The first byte is used for the data-prefix that the display driver requires to be sent when sending pixel data.
+         */ 
+        uint8_t buffer[(wsize.x * wsize.y / 8) + 1] = {0};
```
Comment (LRstudentHU) [inline code/headers/ssd1306_oled_buffered.hpp]: `{0}` could be `{}`

File: code/src/ssd1306.cpp [hunk 11]
```diff
@@ -0,0 +1,46 @@
+#include "ssd1306.hpp"
+
+// ssd1306_i2c_c::
+namespace r2d2::display {
+    ssd1306_i2c_c::ssd1306_i2c_c(r2d2::i2c::i2c_bus_c &bus,
+                                 const uint8_t &address)
+        : bus(bus), address(address), cursor(255, 255) {
+        // wait for the controller to be ready for the initialization
+        hwlib::wait_ms(20);
```
Comment (LRstudentHU) [inline code/src/ssd1306.cpp]: Is this the right place for a wait? This doesn't seem related to the display itself, and the `wait_ms` call has been changed  earlier this PR.

File: code/headers/ssd1306.hpp [hunk 12]
```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <hwlib.hpp>
+#include <i2c_bus.hpp>
+
+namespace r2d2::display {
+    enum class ssd1306_commands : uint8_t {
```
Comment (LRstudentHU) [inline code/headers/ssd1306.hpp]: The enum name should be singular

**Comment by LRstudentHU** [comment]
A couple of nitpicks, after a little cleanup this is ready for merging!

File: code/headers/ssd1306.hpp (hunk 8 above)
Comment (florianhumblot) [reply code/headers/ssd1306.hpp]: These magic numbers are the values of the parameter that is being sent 1 line up. So the 0x80 for example is the clock divider. It wouldn't make sense to add variables for each of them.

File: code/headers/ssd1306.hpp (hunk 9 above)
Comment (florianhumblot) [reply code/headers/ssd1306.hpp]: Not sure how they are inconsistent.

There is a `command` function for sending commands because they need to be prefixed
There is a `data` function for sending data because they need to be prefixed

The only weird function name is `pixels_byte_write` which, while weird, is logical considering it's writing a byte to a pixel. 

I think the names are consistent enough and clear in what they do. Please correct me if I'm wrong.

Comment (LRstudentHU) [reply code/headers/ssd1306.hpp]: `pixels_byte_write` doesn't describe to me what the function is doing, because I can't really imagine what "writing a byte to a pixel" really means. Maybe something like `data_at` or `data_to` would be logical. That's why I suggested the `write_` prefix, so you get `write_data_at` or `write_data_to`.

File: code/src/ssd1306.cpp [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    void ssd1306_i2c_c::pixels_byte_write(hwlib::xy location, uint8_t d) {
+
+        if (location != cursor) {
+            command(ssd1306_commands::column_addr, location.x, 127);
+            command(ssd1306_commands::page_addr, location.y, 7);
+            cursor = location;
+        }
+
+        uint8_t data[] = {ssd1306_data_prefix, d};
+        bus.write(address, data, sizeof(data) / sizeof(uint8_t));
```
Comment (LRstudentHU) [inline code/src/ssd1306.cpp]: You can use the `data` function here.

Comment (florianhumblot) [reply code/src/ssd1306.cpp]: fixed in #12

File: code/headers/ssd1306.hpp (hunk 12 above)
Comment (florianhumblot) [reply code/headers/ssd1306.hpp]: fixed in #12

File: code/src/ssd1306.cpp (hunk 11 above)
Comment (florianhumblot) [reply code/src/ssd1306.cpp]: fixed in #12

File: code/headers/ssd1306_oled_buffered.hpp (hunk 10 above)
Comment (florianhumblot) [reply code/headers/ssd1306_oled_buffered.hpp]: fixed in #12

File: code/headers/ssd1306.hpp (hunk 9 above)
Comment (florianhumblot) [reply code/headers/ssd1306.hpp]: Hmm, I get where you're coming from, but considering it accurately described what it does, and that this is a hardware class not to be used directly, I don't really see the issue.

I'm implementing hwlib::window on the actual instances, which means that the user only sees a `write(xy, color)` in their code, meaning that we get the best of both worlds where I get to be accurate about what the technical function does and the user doesn't need to know about the complexity behind it.

Comment (LRstudentHU) [reply code/headers/ssd1306.hpp]: Fair point. Let's leave it be.
